crates/core/tedge_api/src/messages.rs

```diff
@@ -40,38 +39,173 @@ where
+    pub fn new_with_id(target: &EntityTopicId, cmd_id: String) -> Self {
+        SoftwareListCommand {
+            target: target.clone(),
+            cmd_id,
+            payload: Default::default(),
+        }
     }
-}
 
-impl SoftwareListRequest {
-    pub fn new_with_id(id: &str) -> SoftwareListRequest {
-        SoftwareListRequest { id: id.to_string() }
+    /// Return the MQTT message to register `software-list` support
+    /// TODO: publish the list of supported software package types.
+    pub fn capability_message(schema: &MqttSchema, target: &EntityTopicId) -> Message {
+        let meta_topic = schema.capability_topic_for(target, OperationType::SoftwareList);
+        let payload = "{}";
+        Message::new(&meta_topic, payload)
+            .with_retain()
+            .with_qos(QoS::AtLeastOnce)
     }
 
-    pub fn topic() -> Topic {
-        Topic::new_unchecked(SOFTWARE_LIST_REQUEST_TOPIC)
+    /// Return the SoftwareListCommand received on a topic
+    pub fn try_from(
+        target: EntityTopicId,
+        cmd_id: String,
+        bytes: &[u8],
+    ) -> Result<Option<Self>, serde_json::Error> {
+        if bytes.is_empty() {
+            Ok(None)
+        } else {
+            let payload = SoftwareListCommandPayload::from_slice(bytes)?;
+            Ok(Some(SoftwareListCommand {
+                target,
+                cmd_id,
+                payload,
+            }))
+        }
+    }
+
+    /// Set the status of the command
+    pub fn with_status(mut self, status: CommandStatus) -> Self {
+        self.payload.status = status;
+        self
+    }
+
+    /// Set the failure reason of the command
+    pub fn with_error(mut self, reason: String) -> Self {
+        self.payload.status = CommandStatus::Failed { reason };
+        self
+    }
+
+    /// Add a list of packages all of the same type
+    pub fn add_modules(&mut self, plugin_type: SoftwareType, modules: Vec<SoftwareModule>) {
+        let modules = modules.into_iter().map(|module| module.into()).collect();
+        self.payload.current_software_list.push(SoftwareList {
+            plugin_type,
+            modules,
+        });
+    }
+
+    /// List all the packages
+    pub fn modules(&self) -> Vec<SoftwareModule> {
+        self.payload
+            .current_software_list
+            .iter()
+            .flat_map(|list| {
+                let plugin_type = &list.plugin_type;
+                list.modules
+                    .clone()
+                    .into_iter()
+                    .map(|module| SoftwareModule {
+                        module_type: Some(plugin_type.clone()),
+                        name: module.name,
+                        version: module.version,
+                        url: module.url,
+                        file_path: None,
+                    })
+            })
+            .collect()
+    }
+
+    /// Return the current status of the command
+    pub fn status(&self) -> CommandStatus {
```

Comment:
TODO: find a way (a `trait` extension) to avoid repeating all these methods (`status()`, `topic_id()`, `channel()` ...) for all the command types.

Reply:
Addresses by https://github.com/thin-edge/thin-edge.io/pull/2307/commits/ecc2a84b7883650a1dadb5bce23e1b127db18872